### PR TITLE
Improve compile pipeline

### PR DESCRIPTION
**Pull request by 6543**
## Refactor
- use constants for strings
- more tests
- move constraint code into own package
## Enhance
- all constrains use doublestart (glob pattern matching) now

### CODE REVIEW

File: pipeline/frontend/yaml/constraint/constraint.go [hunk 1]
```diff
@@ -51,20 +50,26 @@ type (
...
 		c.Ref.Match(metadata.Curr.Commit.Ref) &&
 		c.Instance.Match(metadata.Sys.Host) &&
-		c.Matrix.Match(metadata.Job.Matrix) &&
-		c.Path.Match(metadata.Curr.Commit.ChangedFiles, metadata.Curr.Commit.Message)
+		c.Matrix.Match(metadata.Job.Matrix)
+
+	// changed files filter do not apply for tag event
+	if metadata.Curr.Event != frontend.EventTag {
+		match = match && c.Path.Match(metadata.Curr.Commit.ChangedFiles, metadata.Curr.Commit.Message)
+	}
```
Comment (anbraten) [inline pipeline/frontend/yaml/constraint/constraint.go:65]: Not really sure if this is needed as part from the Github tag (push) event no tag event should even set the changedfiles property.

Comment (6543) [reply pipeline/frontend/yaml/constraint/constraint.go:65]: but at the moment we do move that constraint logic into remote - by ensure there no ChangedFiles are set.

but constraint logic should be in one place

Comment (6543) [reply pipeline/frontend/yaml/constraint/constraint.go:65]: tldr: yes we dont need it but it hardens woodpecker & make code more understandable

Comment (anbraten) [reply pipeline/frontend/yaml/constraint/constraint.go:65]: Okay, I see that point. Should we change it to only be active on push and pull-request to be even more specific and safe for future events.

```suggestion
	// changed files filter do not apply for tag event
	if metadata.Curr.Event == frontend.EventPull || metadata.Curr.Event == frontend.EventPush {
		match = match && c.Path.Match(metadata.Curr.Commit.ChangedFiles, metadata.Curr.Commit.Message)
	}
```

Comment (6543) [reply pipeline/frontend/yaml/constraint/constraint.go]: ACK

File: pipeline/frontend/yaml/constraint/constraint.go [hunk 2]
```diff
@@ -51,20 +50,26 @@ type (
...
 		c.Event.Match(metadata.Curr.Event) &&
 		c.Branch.Match(metadata.Curr.Commit.Branch) &&
 		c.Repo.Match(metadata.Repo.Name) &&
 		c.Ref.Match(metadata.Curr.Commit.Ref) &&
 		c.Instance.Match(metadata.Sys.Host) &&
-		c.Matrix.Match(metadata.Job.Matrix) &&
-		c.Path.Match(metadata.Curr.Commit.ChangedFiles, metadata.Curr.Commit.Message)
+		c.Matrix.Match(metadata.Job.Matrix)
+
+	// changed files filter do only apply for pull and push events
```
Comment (anbraten) [inline pipeline/frontend/yaml/constraint/constraint.go:62]:
```suggestion
	// changed files filter do only apply for pull-request and push events
```

**Review by anbraten — APPROVED** [review]

**Pull request merged**